### PR TITLE
Adding max cluster count and size launch test.

### PR DESCRIPTION
**Pull request by kjayapra-amd**
## Motivation

Add cluster occupancy API tests and validate full-scale cluster launch at maximum hardware capacity on gfx1250

## Technical Details

Added hipClusterOccupancy.cc with 4 tests for hipOccupancyMaxActiveClusters/MaxPotentialClusterSize APIs and full-scale launch validation; updated CMakeLists.txt and cluster.yaml

## Issue Tracking
AIRUNTIME-2012

## Test Plan

Four tests: query max active clusters, query max cluster size, launch with max cluster size, launch with both max clusters (128) and max size (16) validating 524K elements

## Test Result

All 4 tests passed on AMD MI450 (gfx1250): 16 max cluster size, 128 max active clusters, 524,288 elements validated successfully

## Submission Checklist

- [X ] Look over the contributing guidelines at https://github.com/ROCm/rocm-systems/blob/develop/CONTRIBUTING.md.

### CODE REVIEW

File: projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+  HIP_CHECK(hipDeviceSynchronize());
+
+  HIP_CHECK(hipMemcpy(hptr_out, dptr_out, num_size, hipMemcpyDeviceToHost));
+
+  REQUIRE(bma.ValidateArrays(hptr_in, hptr_out));
+
+  bma.DestroyHostMemory(hptr_in);
+  bma.DestroyHostMemory(hptr_out);
+  bma.DestroyDeviceMemory(dptr_in);
+  bma.DestroyDeviceMemory(dptr_out);
```
Comment (mangupta) [inline projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Host/device allocations leak when a HIP_CHECK or REQUIRE fails after allocation

In both launch tests the four buffers are released only by the straight-line `bma.Destroy*` calls at the very end of the test body. Every `HIP_CHECK`/`REQUIRE` between allocation and those calls throws out of the test case on failure, so the destroy calls never execute and the host and device allocations are leaked for the remainder of the Catch2 binary.

**Suggested fix:** Make cleanup unconditional: either rely on a RAII owner (let `BasicMemoryAllocator`'s destructor release everything it created and drop the manual `Destroy*` calls), or wrap the buffers in a scope guard / `std::unique_ptr` with a custom deleter so the frees run on the throwing paths as well. Verify the validation result after the resources are released (e.g. capture the bool, free, then `REQUIRE`).

File: projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+ *
+ * Test requirements
+ * ------------------------
+ *  - HIP_VERSION >= 7.0
+ */
+HIP_TEST_CASE(Unit_hipOccupancyMaxActiveClusters_Basic) {
+  if (!CheckTargetSupport()) {
+    INFO("Target Not Supported!");
+    return;
+  }
```
Comment (mangupta) [inline projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Unsupported-target guard reports PASS instead of SKIP and emits no message

All four new test cases guard on `CheckTargetSupport()` with `INFO("Target Not Supported!"); return;`. A Catch2 `INFO` is only flushed when a subsequent assertion in the same scope fails, so on a target without cluster support the test prints nothing and is recorded as a passing test case with zero assertions. Cluster occupancy coverage silently reports green on every device that cannot run it.

**Suggested fix:** Use the skip mechanism the hip-tests guidance mandates at test entry — `HIP_SKIP_TEST("Cluster launch not supported on this target"); return;` — so unsupported configurations are reported as skipped rather than passed, and the reason is actually printed.

File: projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
```
Comment (mangupta) [inline projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** New source file lacks an SPDX-License-Identifier header

The new file carries the full MIT license text but no `SPDX-License-Identifier: MIT` line, which the repository requires on new files for license scanning and package manifest generation.

**Suggested fix:** Add `// SPDX-License-Identifier: MIT` (or the in-comment equivalent) to the header block of the new file, matching the convention used by the other files under `projects/hip-tests/catch/unit/cluster/`.

File: projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+  hipLaunchAttribute attribute[1];
+  attribute[0].id = hipLaunchAttributeClusterDimension;
+  attribute[0].val.clusterDim = {static_cast<unsigned int>(maxClusterSize), 1, 1};
+  config.attrs = attribute;
+  config.numAttrs = 1;
+  config.stream = nullptr;
+  config.dynamicSmemBytes = 0;
+
+  void* kernel_params[] = {&dptr_in, &dptr_out, reinterpret_cast<void*>(const_cast<size_t*>(&num_elems))};
```
Comment (mangupta) [inline projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** `size_t` object passed as the kernel's `int n` argument in the kernel_params array

`ClusterOccupancyKernel` declares its third parameter as `int n`, but the launch passes the address of a `size_t` (`num_elems`) in `kernel_params`. `hipLaunchKernelExC` copies `sizeof(int)` = 4 bytes from that pointer into the argument buffer, so the kernel receives a reinterpretation of half of an 8-byte object rather than a converted value. It happens to yield the right number only because the value is small and the platform is little-endian; there is no conversion and no compile-time type check on the `void*` array.

**Suggested fix:** Introduce an `int` local that matches the kernel parameter — e.g. `int n = static_cast<int>(num_elems);` — and pass `&n` in `kernel_params`, or change the kernel parameter to `size_t n`. In the occupancy-scaled test also assert the element count fits in `int` before narrowing.

File: projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  REQUIRE(maxClusterSize > 0);
+
+  BasicMemoryAllocator<int> bma(num_elems);
+
+  int* hptr_in = bma.CreateAndResetHostMemory();
+  int* hptr_out = bma.CreateAndResetHostMemory();
+  int* dptr_in = bma.CreateAndResetDeviceMemory();
+  int* dptr_out = bma.CreateAndResetDeviceMemory();
+
+  assert(hptr_in != nullptr && hptr_out != nullptr && dptr_in != nullptr && dptr_out != nullptr);
```
Comment (mangupta) [inline projects/hip-tests/catch/unit/cluster/hipClusterOccupancy.cc]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Allocation null checks use `assert`, which is compiled out in Release and aborts the whole binary otherwise

The only validation that the four buffers were actually allocated is a bare `assert(...)`. hip-tests are normally built with `NDEBUG` (Release), where the assert disappears entirely and the following fill loop dereferences a null pointer, crashing the entire Catch2 executable and taking every remaining test case with it. In a debug build the same condition calls `abort()`, which is also a process crash rather than a clean test failure.

**Suggested fix:** Replace the `assert` with Catch2 assertions — `REQUIRE(hptr_in != nullptr);` etc. — so an allocation failure fails the test case cleanly instead of crashing the process, and bound the allocation size in the occupancy-scaled test.

**Review by mangupta — COMMENTED** [review]
🤖 AI-assisted pre-review (for @mangupta)
